sonic-xcvrd/xcvrd/sff_mgr.py

```diff
@@ -0,0 +1,478 @@
+        brought up in a deterministc way, meaning TX is enabled only after
+        host_tx_ready becomes True, and TX will be disabled when host_tx_ready
+        becomes False. This will help eliminate link stability issue and
+        potential interface flap, also turning off TX reduces the power
+        consumption and avoid any lab hazard for admin shut interface.
+
+        Platform can decide whether to enable sff_mgr. By default, it's disabled.
+
+        Pre-requisite for platform to enable sff_mgr:
+        platform needs to keep TX in disabled state after module coming
+        out-of-reset, in either module insertion or bootup cases. This is to
+        make sure the module is not transmitting with TX enabled before
+        host_tx_ready is True.
+        '''
+
+        # CONFIG updates, and STATE_DB for insertion/removal, and host_tx_ready change
+        port_change_observer = PortChangeObserver(self.namespaces,
+                                                  self.logger_for_port_update_event,
+                                                  self.task_stopping_event,
+                                                  self.on_port_update_event,
+                                                  self.PORT_TBL_MAP)
+
+        # This thread doesn't need to expilictly wait on PortInitDone and
+        # PortConfigDone events, as xcvrd main thread waits on them before
+        # spawrning this thread.
+        while not self.task_stopping_event.is_set():
+            # Internally, handle_port_update_event will block for up to
+            # SELECT_TIMEOUT_MSECS until a message is received(in select
+            # function). A message is received when there is a Redis SET/DEL
+            # operation in the DB tables. Upon process restart, messages will be
+            # replayed for all fields, no need to explictly query the DB tables
+            # here.
+            if not port_change_observer.handle_port_update_event():
+                # In the case of no real update, go back to the beginning of the loop
+                continue
+
+            for lport in self.port_dict:
+                if self.task_stopping_event.is_set():
+                    break
+                data = self.port_dict[lport]
+                pport = int(data.get('index', '-1'))
+                subport_idx = int(data.get(self.SUBPORT, '0'))
+                lanes_list = data.get(self.LANES_LIST, None)
+                # active_lanes is a list of boolean values, where True means the
+                # corresponding lane belongs to this logical port.
+                active_lanes = data.get('active_lanes', None)
+                xcvr_type = data.get(self.XCVR_TYPE, None)
+                xcvr_inserted = False
+                host_tx_ready_changed = False
+                admin_status_changed = False
+                if pport < 0 or lanes_list is None:
+                    continue
+
+                if xcvr_type is None:
```

Comment:
@longhuan-cisco please use [_wrapper_get_presence()](https://github.com/sonic-net/sonic-platform-daemons/blob/master/sonic-xcvrd/xcvrd/xcvrd.py#L229). There are cases where SFF manager would be busy doing stuff and may miss the event to process Xcvr removal event.

Reply:
@longhuan-cisco or better use  sfp.get_presence()

Reply:
@prgeor  It's already checked at below:
https://github.com/sonic-net/sonic-platform-daemons/blob/83f7bc3b2dfbbc359380f3b7fbc64127000a1c22/sonic-xcvrd/xcvrd/sff_mgr.py#L401-L406

Reply:
@longhuan-cisco can we remove this redundant check here then because it will be checked further down? The comment says:-

`# TRANSCEIVER_INFO table's XCVR_TYPE is not ready, meaning xcvr is not present`

Reply:
`if xcvr_type is None` is not redundant, because: it is just a simple checking on value already obtained from DB, much more efficient, and in most of time (probably at least >95% of sfp deletion cases, sff_mgr gets to know it via TRANSCEIVER_INFO entry deletion) it can skip the processing earlier than checking of `sfp.get_presence` at later point.